src/v/cloud_storage/materialized_resources.cc

```diff
@@ -706,13 +706,21 @@ class throttled_dl_source : public ss::data_source_impl {
 
 private:
     ss::future<> maybe_throttle(size_t buffer_size) {
-        auto available = _parent._throughput_limit.available();
-        if (available < buffer_size) {
-            // The throttling will be applied. Register the amount.
-            _parent._read_path_probe.download_throttled(
-              buffer_size - available);
+        auto throttled = co_await _parent._throughput_limit.maybe_throttle(
+          buffer_size, _as);
+        if (throttled) {
+            // This path is taken only when the download is throttled. In case
+            // of concurrency every `maybe_throttle` call will return the
+            // precise value for deficiency and sleep time.
+            auto duration
+              = std::chrono::duration_cast<std::chrono::milliseconds>(
+                throttled.value());
+            _parent._read_path_probe.download_throttled(duration.count());
```

Comment:
What are the units for the parameter passed to download_throttled? It looks like this switches from bytes to milliseconds. That's probably worth explaining in the commit message?

Reply:
I updated the description of the metric. It's not possible to reliably capture the metric in bytes (and such metric doesn't make much sense tbh).